core/src/main/java/com/google/cloud/sql/core/DefaultAccessTokenSupplier.java

```diff
@@ -0,0 +1,199 @@
+    retryDuration = Duration.ofSeconds(3);
+  }
+
+  /**
+   * Creates an instance with configurable retry settings.
+   *
+   * @param tokenSource the token source
+   * @param retryCount the number of attempts to refresh.
+   * @param retryDuration the duration to wait between attempts.
+   */
+  @VisibleForTesting
+  DefaultAccessTokenSupplier(
+      Optional<HttpRequestInitializer> tokenSource, int retryCount, Duration retryDuration) {
+    this.tokenSource = tokenSource;
+    this.retryCount = retryCount;
+    this.retryDuration = retryDuration;
+  }
+
+  /**
+   * Parses credentials out of the HttpRequestInitializer.
+   *
+   * @return the GoogleCredentials.
+   * @throws RuntimeException when the HttpRequestInitializer is unrecognized, or if it can't
+   *     produce a GoogleCredentials instance.
+   */
+  private GoogleCredentials parseCredentials() {
+    HttpRequestInitializer source = this.tokenSource.get();
+
+    if (source instanceof HttpCredentialsAdapter) {
+      HttpCredentialsAdapter adapter = (HttpCredentialsAdapter) source;
+      Credentials c = adapter.getCredentials();
+      if (c != null && c instanceof GoogleCredentials) {
+        return (GoogleCredentials) c;
+      }
+      throw new RuntimeException(
+          String.format(
+              "Unable to connect via automatic IAM authentication: "
+                  + "HttpCredentialsAdapter did not create valid credentials. %s, %s",
+              source.getClass().getName(), c));
+    }
+
+    if (source instanceof Credential) {
+      Credential credential = (Credential) source;
+      AccessToken accessToken =
+          new AccessToken(
+              credential.getAccessToken(), getTokenExpirationTime(credential).orElse(null));
+
+      return new GoogleCredentials(accessToken) {
+        @Override
+        public AccessToken refreshAccessToken() throws IOException {
+          credential.refreshToken();
+
+          return new AccessToken(
+              credential.getAccessToken(), getTokenExpirationTime(credential).orElse(null));
+        }
+      };
+    }
+    throw new RuntimeException(
+        String.format(
+            "Unable to connect via automatic IAM authentication: "
+                + "Unsupported credentials of type %s",
+            source.getClass().getName()));
+  }
+
+  /**
+   * Returns an access token, refreshing if the credentials are expired or nearly expired.
+   *
+   * @return the AccessToken or Optional.empty() if there is no tokenSource.
+   * @throws IOException if there is an error attempting to refresh the token
+   * @throws IllegalStateException if the token cannot be used to refresh.
+   */
+  @Override
+  public Optional<AccessToken> get() throws IOException {
+    if (tokenSource.isPresent()) {
+      RetryingCallable<Optional<AccessToken>> retries =
+          new RetryingCallable<>(
+              () -> {
+                final GoogleCredentials credentials;
+                credentials = parseCredentials();
```

Comment:
The combined result of  [CloudSqlInstance:97](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/183c71ab5a6068b1f69d264cd72fbd1fb4fcbc17/core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java#L97) and [SqlAdminApiFetcher.java:301](https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/183c71ab5a6068b1f69d264cd72fbd1fb4fcbc17/core/src/main/java/com/google/cloud/sql/core/SqlAdminApiFetcher.java#L301) with a few important differences: 

1. A lot more error checking.
2. We only refresh the token if it is about to expire (by default `refreshIfExpired()` will make a call to refresh if the token is expired or will expire within the next 5 minutes.